File: Prototype/src/main/java/me/danetnaverno/editoni/minecraft/world/MinecraftRegion.java
package me.danetnaverno.editoni.minecraft.world;

import me.danetnaverno.editoni.common.world.Chunk;
import org.joml.Vector2i;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class MinecraftRegion
{
    public final int x;
    public final int z;

    protected final Map<Vector2i, MinecraftChunk> chunks = new HashMap<>();

    public MinecraftRegion(int x, int z)
    {
        this.x = x;
        this.z = z;
    }

    public Collection<MinecraftChunk> getChunks()
    {
        return new ArrayList<>(chunks.values());
    }

    public void setChunk(MinecraftChunk chunk)
    {
        chunks.put(new Vector2i(chunk.getPosX(), chunk.getPosZ()), chunk);
    }

    public Chunk getChunkByChunkCoord(int chunkX, int chunkZ)
    {
        return chunks.get(new Vector2i(chunkX,chunkZ));
    }

    public Chunk getChunkByBlockCoord(int blockX, int blockZ)
    {
        return getChunkByChunkCoord(blockX >> 4, blockZ >> 4);
    }
}

File: Minecraft/src/main/java/me/danetnaverno/editoni/minecraft/world/MinecraftRegion.kt
package me.danetnaverno.editoni.minecraft.world

import me.danetnaverno.editoni.minecraft.util.location.RegionLocation
import me.danetnaverno.editoni.minecraft.world.io.IMinecraftWorldIOProvider
import me.danetnaverno.editoni.util.location.BlockLocation
import me.danetnaverno.editoni.util.location.ChunkLocation
import java.nio.file.Path
import java.util.*

class MinecraftRegion(@JvmField val world: MinecraftWorld, @JvmField val regionFile: Path, @JvmField val location: RegionLocation)
{
    private val chunks = HashMap<ChunkLocation, MinecraftChunk>()

    fun loadAllChunks()
    {
        (world.worldIOProvider as IMinecraftWorldIOProvider).loadRegion(world, location)
    }

    fun loadChunkAt(chunkLocation: ChunkLocation)
    {
        if (!chunks.containsKey(chunkLocation))
            world.worldIOProvider.loadChunk(world, chunkLocation)
    }

    fun getLoadedChunks(): Collection<MinecraftChunk>
    {
        return chunks.values.toList()
    }

    fun getChunks(): Collection<MinecraftChunk>
    {
        loadAllChunks()
        return chunks.values.toList()
    }

    fun getChunkIfLoaded(location: ChunkLocation): MinecraftChunk?
    {
        return chunks[location]
    }

    fun getChunk(location: ChunkLocation): MinecraftChunk?
    {
        loadChunkAt(location)
        return chunks[location]
    }

    fun getChunk(location: BlockLocation): MinecraftChunk?
    {
        return getChunk(location.toChunkLocation())
    }

    fun setChunk(chunk: MinecraftChunk)
    {
        chunks[chunk.location] = chunk
    }
}